File: problem_3/largest_prime_factor.py
from math import sqrt

def is_prime(factor):
    """Check if a given number is a prime number"""
    for x in range(factor - 2, 1, -2):
        if factor % x == 0: # not a prime
            return False
    return True

def largest_prime_factor(target):
    """Find the largest prime factor of the number given"""
    primes = []
    for i in range(3, int(sqrt(target)), 2):
        if target % i == 0: # is a factor
            if is_prime(i):
                primes.append(i)
    return primes[-1]

number = 600851475143
print(f"The largest prime factor of {number} is {largest_prime_factor(number)}")

File: problem_3/largest_prime_factor.rb
# frozen_string_literal: true

def prime?(factor)
  (factor - 2...1).step(-2).each do |x|
    return false if (factor % x).zero?
  end
  true
end

def largest_prime_factor(target)
  primes = []
  (3..Integer.sqrt(target)).step(2).each do |i|
    if (target % i).zero?
      primes.push(i) if prime?(i)
    end
  end
  primes.pop
end

number = 600_851_475_143
puts "The largets prime factor of #{number} is #{largest_prime_factor(number)}"
